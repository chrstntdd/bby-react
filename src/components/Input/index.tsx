import React, { PureComponent, Fragment } from 'react';

import { noop, debounce } from '@/util';
import { Maybe } from '@/fp';

interface PInput {
  name: string;
  inputRef: any;
  validateFn?: (val) => Maybe<string>;
  validationCb?: (inputId: any, isValid: boolean, value?: string) => void;
}

interface SInput {
  value: string;
  validationMsg: string;
}

export class Input extends PureComponent<PInput & React.HTMLProps<HTMLInputElement>, SInput> {
  public static defaultProps: Partial<PInput> = {
    validateFn: noop,
    validationCb: noop
  };

  state = {
    value: '',
    validationMsg: ''
  };

  handleValidation = debounce((inputValue: string) => {
    const validationMsg = this.props.validateFn(inputValue);

    validationMsg.caseOf({
      just: msg => {
        /* FIELD IS INVALID */
        this.props.validationCb(this.props.name, false, inputValue);

        this.setState({ validationMsg: msg });
      },

      nothing: () => {
        /* FIELD IS VALID */
        this.props.validationCb(this.props.name, true, inputValue);

        this.setState({ validationMsg: '' });
      }
    });
  }, 200);

  handleChange = (e: React.FormEvent<HTMLInputElement>) => {
    const inputValue = e.currentTarget.value;

    this.setState({ value: inputValue }, () => this.handleValidation(inputValue));
  };

  render() {
    /* THOUGH THE VALIDATION FUNCTIONS ARE NOT USED HERE, BUT IN ORDER TO KEEP 
       REACT HAPPY, THE EXTRA PROPS SHOULD NOT BE ADDED TO THE INPUT ELEMENT
    */
    const { name, label, inputRef, validateFn, validationCb, ...rest } = this.props;
    const { validationMsg, value } = this.state;

    return (
      <Fragment>
        {validationMsg ? <span className="error-span">{validationMsg}</span> : null}
        <input
          {...rest}
          ref={inputRef}
          onChange={this.handleChange}
          className={value === '' ? '' : 'has-content'}
        />
        {label ? <label htmlFor={name}>{label}</label> : null}
      </Fragment>
    );
  }
}

export default Input;
