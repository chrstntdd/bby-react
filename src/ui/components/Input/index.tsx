import React, { Fragment } from 'react';

interface PInput {
  id: string;
  inputRef: any;
  label: string;
  validationMsg?: string;
  onChange?: (e) => void;
}

const Input: React.SFC<PInput & React.HTMLProps<HTMLInputElement>> = ({
  id,
  label,
  inputRef,
  onChange,
  validationMsg,
  ...inputAttrs
}) => {
  return (
    <Fragment>
      {<label htmlFor={id}>{label}</label>}
      <input {...inputAttrs} autoComplete="off" id={id} ref={inputRef} onChange={onChange} />
      <span className="error-span">{validationMsg}</span>
    </Fragment>
  );
};

export default Input;
