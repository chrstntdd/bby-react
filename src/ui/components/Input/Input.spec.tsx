import React from 'react';
import { shallow } from 'enzyme';

import Input from './';

const requiredProps = {
  label: 'test',
  id: '1234',
  inputRef: null,
  validationMsg: ''
};

describe('Input component', () => {
  it('should render', () => {
    const wrapper = shallow(<Input {...requiredProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call onChange prop callback if provided', () => {
    const onChangeMock = jest.fn();
    const mockEvent = { target: { value: 'test string' } };
    const wrapper = shallow(<Input {...requiredProps} onChange={onChangeMock} />);

    wrapper.find('input').simulate('change', mockEvent);

    expect(onChangeMock).toHaveBeenCalledTimes(1);
    expect(onChangeMock).toHaveBeenCalledWith(mockEvent);
  });
});
