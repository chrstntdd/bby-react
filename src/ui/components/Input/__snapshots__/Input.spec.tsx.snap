// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render 1`] = `
<React.Fragment>
  <label
    htmlFor="1234"
  >
    test
  </label>
  <input
    autoComplete="off"
    id="1234"
  />
  <span
    className="error-span"
  />
</React.Fragment>
`;
